### PR TITLE
feat: initial commit of `wdl-docs` style

### PR DESCRIPTION
**Pull request by claymcleod**
This PR adds a basic style scaffold for producing the CSS file used in `wdl-doc`. Styles can be added in the `src` folder and compiled into development (`pnpm run dev`, watches the directory and serves the `example.html` file live) or production (`pnpm run prod`, which needs a subsequent `pnpm run serve` to serve the `example.html` file).

You can update the `example.html` file with a fully-featured generated page that we can test with.

- Please use `pnpm`, as it's way better than `npm` in my experience.

Before submitting this PR, please make sure:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [x] You have added at least one relevant code reviewer to the PR.
- [ ] Your code builds clean without any errors or warnings.
- [ ] You have added an entry to the relevant `CHANGELOG.md` (see
      ["keep a changelog"] for more information).
- [x] Your commit messages follow the [conventional commit] style.

[conventional commit]: https://www.conventionalcommits.org/en/v1.0.0/#summary
["keep a changelog"]: https://keepachangelog.com/en/1.0.0/

### CODE REVIEW

File: wdl-doc/tests/output_docs/data_structures/read_group/read_group_to_string-task.html [hunk 1]
```diff
@@ -1,6 +1,3 @@
-<h1>read_group_to_string</h1><ul><li>Meta:meta {
-        description: "Stringifies a ReadGroup struct"
-        outputs: {
-            stringified_read_group: "Input ReadGroup as a string"
-        }
-    }</li><li>Inputs:<ul><li><h2>read_group</h2><p>Type: ReadGroup</p><p>Expr: None</p></li></ul></li><li>Outputs:<ul><li><h2>stringified_read_group</h2><p>Type: String</p><p>Expr: read_string("out.txt")</p></li></ul></li></ul>
\ No newline at end of file
+<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><title>read_group_to_string</title><link rel="stylesheet" type="text/css" href="/Users/afrantz/wdl/wdl-doc/theme/dist/styles.css"></head></html><h1>read_group_to_string</h1><h2>Meta</h2><ul><li><b>description:</b> &quot;Stringifies a ReadGroup struct&quot;</li><li><b>outputs:</b> {
```
Comment (claymcleod) [inline wdl-doc/tests/output_docs/data_structures/read_group/read_group_to_string-task.html:1]: 1. We should remove the references to your home directory somehow.
2. It looks like the `<html>` tag closes before your elements are added in.

Comment (a-frantz) [reply wdl-doc/tests/output_docs/data_structures/read_group/read_group_to_string-task.html:1]: 1. Already got a WIP addressing that by using a relative path instead (though it complicates things a touch). I noticed that right after I made the WIP commit.
2. I'll look into this!

File: wdl-doc/src/docs_tree.rs [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        &self.path
+    }
+
+    /// Get the page associated with the node.
+    pub fn page(&self) -> Option<Rc<HTMLPage>> {
+        self.page.clone()
+    }
+
+    /// Iterate over the children of the node in a Depth First Traversal order.
+    pub fn depth_first_traversal(&self) -> Vec<&Node> {
```
Comment (peterhuene) [inline wdl-doc/src/docs_tree.rs:98]: I think this function could be implemented with a local function that takes the `Vec` by mutable reference so that it doesn't have to allocate and deallocate a `Vec` at every node.

For example:

```rust
pub fn depth_first_traversal(&self) -> Vec<&Node> {
    fn recurse_depth_first(node: &Node, nodes: &mut Vec<&Node>) {
        nodes.push(self);

        for child in self.children.values() {
            recurse_depth_first(child, nodes);
        }
    }

    let mut nodes = Vec::new();
    recurse_depth_first(self, &mut nodes);

    nodes
}
```

Comment (peterhuene) [reply wdl-doc/src/docs_tree.rs:98]: Alternatively, depending on how this method is used, we could not return a `Vec` and instead accept a `FnMut` callback that would "visit" each node in DFS order.

It certainly looks like the places we use this method could be translated to:

```rust
root.depth_first_traversal(|node| {
    // do something with node
});
```

File: wdl-doc/src/docs_tree.rs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    ///
+    /// `root.path` is the path to the DOCS directory and is absolute.
+    root: Node,
+    /// The absolute path to the stylesheet, if it exists.
+    stylesheet: Option<PathBuf>,
+}
+
+impl DocsTree {
+    /// Create a new DOCS tree.
+    pub fn new<P: AsRef<Path>>(root: P, stylesheet_to_copy: Option<P>) -> Self {
```
Comment (peterhuene) [inline wdl-doc/src/docs_tree.rs:123]: When repeating a type parameter in a signature like this, it's probably best to use `impl AsRef<Path>` in the signatures so that they don't have to be the same type (i.e. one should be able to call this with `new("foo", Some(Path::new("foo")))`. It's functionally equivalent to having two type parameters `P: AsRef<Path>, Q: AsRef<Path>`.

However, the `Option<impl AsRef<Path>>` is annoying to deal with as you would have to qualify it in the `None` case.

When I see something like that, I tend to break the method into two: `new(root: impl AsRef<Path>)` and `new_with_stylesheet(root: impl AsRef<Path>, stylesheet: impl AsRef<Path>)` and let it internally deal with the `Option`.

File: wdl-doc/src/callable/task.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        parameter_meta: Option<ParameterMetadataSection>,
+        input_section: Option<InputSection>,
+        output_section: Option<OutputSection>,
+        runtime_section: Option<RuntimeSection>,
+    ) -> Self {
+        let meta = if let Some(ref mds) = meta_section {
+            parse_meta(mds)
+        } else {
+            MetaMap::default()
+        };
```
Comment (peterhuene) [inline wdl-doc/src/callable/task.rs:44]: Nit for consistency since the other `if let` expressions don't use `ref`:

```suggestion
        let meta = if let Some(mds) = meta_section {
            parse_meta(&mds)
        } else {
            MetaMap::default()
        };
```

File: wdl-doc/src/callable/workflow.rs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        meta_section: Option<MetadataSection>,
+        parameter_meta: Option<ParameterMetadataSection>,
+        input_section: Option<InputSection>,
+        output_section: Option<OutputSection>,
+    ) -> Self {
+        let meta = if let Some(ref mds) = meta_section {
+            parse_meta(mds)
+        } else {
+            MetaMap::default()
+        };
```
Comment (peterhuene) [inline wdl-doc/src/callable/workflow.rs:41]: Same comment as above re: `let` usage:

```suggestion
        let meta = if let Some(mds) = meta_section {
            parse_meta(&mds)
        } else {
            MetaMap::default()
        };
```

File: wdl-doc/src/docs_tree.rs [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+
+        let html = full_page(
+            page.name(),
+            html! {
+                (sidebar)
+                (content)
+            },
+            stylesheet.as_deref(),
+        );
+        std::fs::write(path, html.into_string()).unwrap();
```
Comment (peterhuene) [inline wdl-doc/src/docs_tree.rs:334]: We should probably functions that interact with the file system fallible rather than panic on IO error.

File: wdl-doc/src/docs_tree.rs (hunk 2 above)
Comment (peterhuene) [reply wdl-doc/src/docs_tree.rs:98]: Although, looking at what `maud` supports, we probably can't use this within a `html!` macro, so returning a `Vec` is fine.

**Comment by peterhuene** [comment]
I was going to message this on Slack, but apparently Slack is having an outage:

> I put some minor comments on the PR, although I know it's not done yet; generally I think the approach looks really good!

File: wdl-doc/src/docs_tree.rs (hunk 3 above)
Comment (a-frantz) [reply wdl-doc/src/docs_tree.rs]: Would ask over Slack, but it's still down! I've updated `DocsTree::new` to just take an `impl AsRef<Path>` and then wrote `new_with_stylesheet` like you usggested.

The pattern you pointed here is also in use for the `document_workspace()` entrypoint function, but I'm not sure the right way to refactor that. Should there be a pair of `document_worskspace()` and `document_workspace_with_stylesheet()` functions? Or is there another way to refactor it that you would recommend?

File: wdl-doc/src/docs_tree.rs (hunk 6 above)
Comment (a-frantz) [reply wdl-doc/src/docs_tree.rs]: good point! I pretty heavily use `unwrap()` in this PR as a shortcut, but just went through and updated some functions to return `anyhow::Result` instead and also changed some unwraps to expects.

File: wdl-doc/src/docs_tree.rs (hunk 3 above)
Comment (peterhuene) [reply wdl-doc/src/docs_tree.rs]: `document_workspace_with_stylesheet` definitely is a mouthful; if we want two different functions, I'm sure we'd all prefer something less verbose.

Still, perhaps having the `Option` parameter is fine and folks won't mind passing something like `None::<&str>`? I don't have any suggested alternatives, but I still think we should be using different type parameters at the very least.

File: wdl-doc/src/callable.rs [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+                    tbody class="border" {
+                        @for param in iter {
+                            (param.render())
+                        }
+                    }
+                }
+            };
+        };
+        html! {}
+    }
```
Comment (adthrasher) [inline wdl-doc/src/callable.rs:134]: Do we need to treat `Common` separately? Or can we apply a custom sort to `input_groups`?

File: wdl-doc/src/callable.rs [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+                        (param.render())
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Parse a [`MetadataSection`] into a [`MetaMap`].
+fn parse_meta(meta: &MetadataSection) -> MetaMap {
```
Comment (adthrasher) [inline wdl-doc/src/callable.rs:227]: Can we process `meta` and `parameter_meta` together? Is there a generic trait or type that both share to enable a single function?

Comment (a-frantz) [reply wdl-doc/src/callable.rs:227]: I can play around with this locally and see if I can join these methods together. Good thought

File: wdl-doc/src/callable.rs (hunk 7 above)
Comment (a-frantz) [reply wdl-doc/src/callable.rs:134]: That should be possible. Another good thought.
I think if we're going down a custom sort order for `input_groupts`, `Common` should come first, and `Resources` should come last? Everything else would just be alphabetical between those two?

File: wdl-doc/src/callable.rs (hunk 8 above)
Comment (a-frantz) [reply wdl-doc/src/callable.rs:230]: Actually this isn't so straightforward. I played around with it for a bit locally, and ran into other issues. I think we have to do it like this 🤷

File: wdl-doc/src/callable.rs [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+    fn input_groups(&self) -> BTreeSet<Group> {
+        self.inputs()
+            .iter()
+            .filter_map(|param| param.group())
+            .map(|arg0: Group| Group(arg0.0.clone()))
+            .collect()
+    }
+
+    /// Get the inputs of the callable that are part of `group`.
+    fn inputs_in_group<'a>(&'a self, group: &'a Group) -> impl Iterator<Item = &'a Parameter> {
```
Comment (claymcleod) [inline wdl-doc/src/callable.rs:98]: For this and the next method, there's actually a cool method I wasn't aware of called [`partition()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition) that can partition the inputs into "grouped" or "non-grouped" in one call. I'd consider using that.

Comment (claymcleod) [reply wdl-doc/src/callable.rs:98]: I'm unclear if this is easy with what's going on with required. So it's more just a thing to consider.

File: wdl-doc/src/callable.rs [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+            let name = decl.name().as_str().to_owned();
+            let meta = parameter_meta.get(&name).or_else(|| output_meta.get(&name));
+            Parameter::new(
+                wdl_ast::v1::Decl::Bound(decl.clone()),
+                meta.cloned(),
+                InputOutput::Output,
+            )
+        })
+        .collect()
+}
```
Comment (claymcleod) [inline wdl-doc/src/callable.rs:295]: Is there a way to add some simple tests for this?

File: wdl-doc/src/callable/task.rs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    fn inputs(&self) -> &[Parameter] {
+        &self.inputs
+    }
+
+    fn outputs(&self) -> &[Parameter] {
+        &self.outputs
+    }
+}
```
Comment (claymcleod) [inline wdl-doc/src/callable/task.rs:148]: Needs simple tests.

File: wdl-doc/src/callable/workflow.rs [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    fn inputs(&self) -> &[Parameter] {
+        &self.inputs
+    }
+
+    fn outputs(&self) -> &[Parameter] {
+        &self.outputs
+    }
+}
```
Comment (claymcleod) [inline wdl-doc/src/callable/workflow.rs:136]: Needs simple tests.

File: wdl-doc/src/docs_tree.rs [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+        &self.name
+    }
+
+    /// Get the type of the page.
+    pub fn page_type(&self) -> &PageType {
+        &self.page_type
+    }
+}
+
+/// A node in the DOCS directory tree.
```
Comment (claymcleod) [inline wdl-doc/src/docs_tree.rs:58]:
```suggestion
/// A node in the docs directory tree.
```

File: wdl-doc/src/docs_tree.rs [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+    /// The root of the tree.
+    ///
+    /// `root.path` is the path to the DOCS directory and is absolute.
+    root: Node,
+    /// The absolute path to the stylesheet, if it exists.
+    stylesheet: Option<PathBuf>,
+}
+
+impl DocsTree {
+    /// Create a new DOCS tree.
```
Comment (claymcleod) [inline wdl-doc/src/docs_tree.rs:126]:
```suggestion
    /// Create a new docs tree.
```

File: wdl-doc/src/docs_tree.rs [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+        }
+
+        let mut nodes = Vec::new();
+        recurse_depth_first(self, &mut nodes);
+
+        nodes
+    }
+}
+
+/// A tree representing the DOCS directory.
```
Comment (claymcleod) [inline wdl-doc/src/docs_tree.rs:114]: The `DOCS` thing needs to be universally searched for and replaced—seems like an errant find and replace maybe.

File: wdl-doc/src/lib.rs [hunk 16]
```diff
@@ -6,59 +6,162 @@
...
+        let mut options = Options::empty();
+        options.insert(Options::ENABLE_TABLES);
+        options.insert(Options::ENABLE_STRIKETHROUGH);
+        let parser = Parser::new_ext(self.0.as_ref(), options);
+        pulldown_cmark::html::push_html(&mut unsafe_html, parser);
+        // Sanitize it with ammonia
+        let safe_html = ammonia::clean(&unsafe_html);
+
+        // Remove the outer `<p>` tag that `pulldown_cmark` wraps single lines in
+        let safe_html = if safe_html.starts_with("<p>") && safe_html.ends_with("</p>\n") {
```
Comment (claymcleod) [inline wdl-doc/src/lib.rs:111]: This needs a test, both the positive and negative case.

File: wdl-doc/src/lib.rs [hunk 17]
```diff
@@ -67,282 +170,198 @@ impl Document {
...
+                        }}
+                        tbody class="border" {
+                            @for page in &self.local_pages {
+                                tr class="border" {
+                                    td class="border" {
+                                        a href=(page.0.to_str().unwrap()) { (page.1.name()) }
+                                    }
+                                    td class="border" {
+                                        @match page.1.page_type() {
+                                            PageType::Index(_) => { "TODO ERROR" }
```
Comment (claymcleod) [inline wdl-doc/src/lib.rs:207]: Can this not be a literal `todo!()`?

File: wdl-doc/src/lib.rs [hunk 18]
```diff
@@ -67,282 +170,198 @@ impl Document {
...
-            .file_name()
-            .expect("current directory should have a file name")
-            .to_string_lossy();
         let ast_doc = result.document().node();
         let version = ast_doc
             .version_statement()
-            .expect("document should have a version statement")
-            .version();
-        let preamble = fetch_preamble_comments(ast_doc.clone());
+            .expect("Document should have a version statement");
```
Comment (claymcleod) [inline wdl-doc/src/lib.rs:288]:
```suggestion
            .expect("document should have a version statement");
```

These expect statements should be all lowercase to follow the existing convention (unless you're starting with a proper noun, like `WDL`).

Comment (claymcleod) [inline wdl-doc/src/lib.rs:288]:
```suggestion
            .expect("document should have a version statement");
```

These expect statements should be all lowercase to follow the existing convention (unless you're starting with a proper noun, like `WDL`).

Comment (claymcleod) [reply wdl-doc/src/lib.rs:288]: Please double check all the expect statements here.

File: wdl-doc/src/lib.rs [hunk 19]
```diff
@@ -67,282 +170,198 @@ impl Document {
...
+        let mut local_pages = Vec::new();
 
         for item in ast.items() {
             match item {
                 DocumentItem::Struct(s) => {
-                    let struct_name = s.name().as_str().to_owned();
-                    let struct_file = cur_dir.join(format!("{}-struct.html", struct_name));
-                    let mut struct_file = tokio::fs::File::create(struct_file).await?;
+                    let name = s.name().as_str().to_owned();
+                    let path = cur_dir.join(format!("{}-struct.html", name));
```
Comment (claymcleod) [inline wdl-doc/src/lib.rs:297]: It might be a minor thing, but I would advocate for `struct-{}.html`, since that more closely follows the hierarchy here.

**Review by claymcleod — COMMENTED** [review]
I can't approve the PR, as I'm the one who made it, but I'm generally good with this provided these are addressed.

File: wdl-doc/src/lib.rs (hunk 17 above)
Comment (a-frantz) [reply wdl-doc/src/lib.rs:207]: Unfortunately, that doesn't work within the `maud` macro. Trying to write `unreachable!()` or other such panic gets the entire macro triggered as "unreachable code block" by clippy. I saw some folks online using those panics, but they were also silencing the clippy warning.

File: wdl-doc/src/lib.rs (hunk 19 above)
Comment (claymcleod) [reply wdl-doc/src/lib.rs:297]: Okay that's fine I suppose. I definitely think that classifying it into buckets of struct/workflow/task is a higher-order ranking of the objects. But it would probably make more sense to do `/struct/<name`, `/task/<name>`, and `/workflow/<name>` in that case.

Comment (adthrasher) [reply wdl-doc/src/lib.rs:305]: Somewhat related, I can see some repositories wanting the sidebar to look like:

- structs
-- struct 1
-- struct 2
- tasks
-- task 1
-- task 2

Rather than being grouped by source file. Clay's naming convention would make that easy to implement.

**Review by adthrasher — APPROVED** [review]

**Comment by peterhuene** [comment]
Ugh. I need to look into authenticating CI so we don't hit the rate limit.

Also, probably need to implement caching of the check for image presence in crankshaft so that we don't unnecessarily query for the image again for the same run after ensuring it is present.
